Source/GSOrderedSet.m

```diff
@@ -0,0 +1,332 @@
+/** Concrete implementation of GSOrderedSet and GSMutableOrderedSet
+    based on GNU NSOrderedSet and NSMutableOrderedSet classes
+    Copyright (C) 2019 Free Software Foundation, Inc.
+    
+    Written by: Gregory Casamento <greg.casamento@gmail.com>
+    Created: May 17 2019
+    
+    This file is part of the GNUstep Base Library.
+    
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    _version 2 of the License, or (at your option) any later _version.
+    
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Library General Public License for more details.
+    
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free
+    Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+    Boston, MA 02111 USA.
+*/
+
+#import "common.h"
+#import "Foundation/NSOrderedSet.h"
+#import "Foundation/NSAutoreleasePool.h"
+#import "Foundation/NSArray.h"
+#import "Foundation/NSEnumerator.h"
+#import "Foundation/NSException.h"
+#import "Foundation/NSPortCoder.h"
+#import "Foundation/NSIndexSet.h"
+#import "Foundation/NSKeyedArchiver.h"
+#import "Foundation/NSValue.h"
+#import "Foundation/NSDictionary.h"
+#import "GNUstepBase/GSObjCRuntime.h"
+#import "GSPrivate.h"
+#import "GSFastEnumeration.h"
+#import "GSDispatch.h"
+#import "GSSorting.h"
+
+#define	GSI_ARRAY_TYPE	NSRange
```

Comment:
OK, here is your problem with GSIArray: Whenever GSI_ARRAY_NO_RETAIN is defined, no matter to which value, there won't be any retain code called. What you need to do is remove all the following six lines, just leave the one with
#define GSI_ARRAY_TYPES       GSUNION_OBJ

Next you remove the retain calls from the code and as I tested locally your tests all still work.